wa/workloads/speedometer/__init__.py

```diff
@@ -12,57 +12,221 @@
+        )
+
+        webserver_cmd = "cd {} && python3 -mhttp.server {}".format(
+            self.document_root, self.webserver_port
+        )
+
+        with open(os.devnull, "w") as dev_null:
+            self.webserver = subprocess.Popen(
+                webserver_cmd, shell=True, stdout=dev_null, stderr=dev_null
+            )
+
+        subprocess.check_output(
+            "adb reverse tcp:{0} tcp:{0}".format(self.webserver_port), shell=True
+        )
+
+        # Clear the cache...
+        self.target.execute("pm clear {}".format(self.chrome_package, as_root=True))
+
+        # Launch the browser for the first time...
+        browser_launch_cmd = "am start -a android.intent.action.VIEW -d {} {}".format(
+            self.speedometer_url, self.chrome_package
+        )
+
+        self.target.execute(browser_launch_cmd)
+        time.sleep(1)
+        self.target.execute("am force-stop {}".format(self.chrome_package))
+        time.sleep(1)
+
+        # Pull the preferences file from the device, modify it, and push it back.
+        # This is done to bypass the 'first launch' screen of the browser.
+        self.preferences_xml = "{}_preferences.xml".format(self.chrome_package)
+
+        self.target.pull(
+            "/data/data/{}/shared_prefs/{}".format(
+                self.chrome_package, self.preferences_xml
+            ),
+            context.output_directory,
+            as_root=True,
+        )
+
+        with open(
+            os.path.join(context.output_directory, self.preferences_xml)
+        ) as read_fh:
+            lines = read_fh.readlines()
+            lines.insert(
+                len(lines) - 1, '<boolean name="first_run_flow" value="true" />\n'
+            )
+            lines.insert(
+                len(lines) - 1,
+                '<boolean name="first_run_tos_accepted" value="true" />\n',
+            )
+            lines.insert(
+                len(lines) - 1,
+                '<boolean name="first_run_signin_complete" value="true" />\n',
+            )
+            with open(
+                os.path.join(context.output_directory, self.preferences_xml + ".new"),
+                "w",
+            ) as write_fh:
+                for line in lines:
+                    write_fh.write(line)
+
+        self.target.push(
+            os.path.join(context.output_directory, self.preferences_xml + ".new"),
+            "/data/data/{}/shared_prefs/{}".format(
+                self.chrome_package, self.preferences_xml
+            ),
+            as_root=True,
+        )
+
+    def run(self, context):
+        super(Speedometer, self).run(context)
+
+        browser_launch_cmd = "am start -a android.intent.action.VIEW -d {} {}".format(
+            self.speedometer_url, self.chrome_package
+        )
+        self.target.execute(browser_launch_cmd)
+
+        # Wait 60 seconds at least, and then wait until we don't see the 'sandboxed_process' process for 10 seconds.
+        time.sleep(60)
```

Comment:
Why sixty? You should avoid long sleeps if _at all_ possible. Can you not poll running packages or logcat instead?

Reply:
The idea is to reduce the amount of polling we do to the system while it's supposed to be running a performance-sensitive benchmark. Since we inevitably are going to need to poll in some way to work out which it has finished, a compromise was to wait at least 60 seconds - I know Speedometer usually takes around 90 seconds on current devices. Is there something intrinsically wrong with sleeping for this long? What does it affect?

Reply:
Long sleeps introduce unnecessary delays, this can end up significantly increasing the over all time. Yo're also making assumptions about how long things take which may not be valid when run on a system that is vastly different from the system on which the workload is first implemented. Fixed waits are brittle. 

This is why, as a rule, we generally prefer some form of polling or signalling guarded by generous timeouts.   A slow low-overhead poll (e.g. with cadence of 1-2s) will typically not impact most workloads significantly enough to affect results -- given the level of backgrounds activity on a typical rich-OS system (especially Android), it's usually in the noise.